### PR TITLE
[A8] - Software and Data Integrity Failures - Python - Amarelo Designs - Solution

### PR DESCRIPTION
**Pull request by henriporto**
## This solution refers to which of the apps?

[A8 - Software and Data Integrity Failures - Python - Amarelo Designs](https://github.com/globocom/secDevLabs/tree/master/owasp-top10-2021-apps/a8/amarelo-designs)

## What did you do to mitigate the vulnerability?

Removed pickle library and implemented session (properly signed).

## Did you test your changes? What commands did you run?

Yes. I followed the attack narrative.

### CODE REVIEW

File: owasp-top10-2021-apps/a8/amarelo-designs/app/app.py [hunk 1]
```diff
@@ -19,12 +23,9 @@ def login():
     
         if username == "admin" and password == "admin":
```
Comment (gustavocovas) [inline owasp-top10-2021-apps/a8/amarelo-designs/app/app.py:24]: What about these hardcoded credentials?

File: owasp-top10-2021-apps/a8/amarelo-designs/app/app.py [hunk 2]
```diff
@@ -1,11 +1,15 @@
 # coding: utf-8
 
-from flask import Flask, request, make_response, render_template, redirect, flash
+from flask import Flask, request, make_response, render_template, redirect, session
```
Comment (gustavocovas) [inline owasp-top10-2021-apps/a8/amarelo-designs/app/app.py:3]: Cool, great choice of library

File: owasp-top10-2021-apps/a8/amarelo-designs/app/app.py [hunk 3]
```diff
@@ -19,12 +23,9 @@ def login():
     
         if username == "admin" and password == "admin":
             token = str(uuid.uuid4().hex)
-            cookie = { "username":username, "admin":True, "sessionId":token }
-            pickle_resultado = pickle.dumps(cookie)
-            encodedSessionCookie = base64.b64encode(pickle_resultado)
-            resp = make_response(redirect("/user"))
-            resp.set_cookie("sessionId", encodedSessionCookie)
-            return resp
+            admin_token.append(token)
```
Comment (gustavocovas) [inline owasp-top10-2021-apps/a8/amarelo-designs/app/app.py:26]: Admin tokens are stored in a local variable. But are they removed when expired?
What would happen if we have a HA setup with 2 instances of this app? Would an user authenticated as admin in the first one be recognized as such in the second one?

**Review by gustavocovas — CHANGES_REQUESTED** [review]

File: owasp-top10-2021-apps/a8/amarelo-designs/app/app.py (hunk 1 above)
Comment (henriporto) [reply owasp-top10-2021-apps/a8/amarelo-designs/app/app.py]: fixed!

File: owasp-top10-2021-apps/a8/amarelo-designs/app/app.py (hunk 2 above)
Comment (henriporto) [reply owasp-top10-2021-apps/a8/amarelo-designs/app/app.py]: xD

File: owasp-top10-2021-apps/a8/amarelo-designs/app/app.py (hunk 3 above)
Comment (henriporto) [reply owasp-top10-2021-apps/a8/amarelo-designs/app/app.py]: It was fixed with:
mariadb + event scheduler + logging.

Captura de Tela 2022-04-14 às 13 53 37

**Comment by gustavocovas** [comment]
Hey, @henriporto! 

MariaDB is a great choice for a relational database, and you implementation looks good and within the standards for SecDevLabs. However, as we discussed in private, relational databases may be not the best solution for event scheduling.

I would suggest that you take a look at using [Flask Sessions with Redis](https://flask-session.readthedocs.io/en/latest/#redissessioninterface).

Thank you and keep up the great work! 🚀

**Comment by henriporto** [comment]
> Hey, @henriporto!
> 
> MariaDB is a great choice for a relational database, and you implementation looks good and within the standards for SecDevLabs. However, as we discussed in private, relational databases may be not the best solution for event scheduling.
> 
> I would suggest that you take a look at using [Flask Sessions with Redis](https://flask-session.readthedocs.io/en/latest/#redissessioninterface).
> 
> Thank you and keep up the great work! 🚀

Thanks! Fixed with MariaDB, JWT, Argon2ID.

**Pull request closed without merging**